src/mesh/distributed_mesh.C

```diff
@@ -1442,12 +1457,133 @@ void DistributedMesh::renumber_nodes_and_elements ()
 
   LOG_SCOPE("renumber_nodes_and_elements()", "DistributedMesh");
 
+  // Nodes not connected to any elements, and nullptr node entries
+  // in our container, should be deleted.  But wait!  If we've deleted coarse
+  // local elements on some processor, other processors might have ghosted
+  // nodes from it that are now no longer connected to any elements on it, but
+  // that are connected to their own semilocal elements.  We'll have to
+  // communicate to ascertain if that's the case.
+  this->comm().max(_deleted_coarse_elements);
+
+  // What used nodes do we see on our proc?
   std::set<dof_id_type> used_nodes;
 
+  // What used node info should we send from our proc?  Could we take ownership
+  // of each if we needed to?
+  std::map<processor_id_type, std::map<dof_id_type, bool>>
+    used_nodes_on_proc;
+
   // flag the nodes we need
   for (auto & elem : this->element_ptr_range())
     for (const Node & node : elem->node_ref_range())
-      used_nodes.insert(node.id());
+      {
+        const dof_id_type n = node.id();
+        used_nodes.insert(n);
+        if (_deleted_coarse_elements)
+          {
+            const processor_id_type p = node.processor_id();
+            if (p != this->processor_id())
+              {
+                auto & used_nodes_on_p = used_nodes_on_proc[p];
+                if (elem->processor_id() == this->processor_id())
+                  used_nodes_on_p[n] = true;
+                else
+                  if (!used_nodes_on_p.count(n))
+                    used_nodes_on_p[n] = false;
+              }
+          }
+      }
+
+  if (_deleted_coarse_elements)
+    {
+      // "unsigned char" == "bool, but MPI::BOOL is iffy to use"
+      typedef unsigned char boolish;
+      std::map<processor_id_type, std::vector<std::pair<dof_id_type, boolish>>>
+        used_nodes_on_proc_vecs;
+      for (auto & [pid, nodemap] : used_nodes_on_proc)
+        used_nodes_on_proc_vecs[pid].assign(nodemap.begin(), nodemap.end());
+
+      std::map<dof_id_type,processor_id_type> repartitioned_node_pids;
+      std::map<processor_id_type, std::set<dof_id_type>>
+        repartitioned_node_sets_to_push;
+
+      auto ids_action_functor =
+        [this, &used_nodes, &repartitioned_node_pids,
+         &repartitioned_node_sets_to_push]
+        (processor_id_type pid,
+         const std::vector<std::pair<dof_id_type, boolish>> & ids_and_bools)
+        {
+          for (auto [n, local_to_pid] : ids_and_bools)
```

Comment:
I would rename `local_to_pid` to something like `can_make_owning_pid` or `can_make_sender_owner`. I confused myself for a while thinking that `local_to_pid` was a `processor_id_type`

Reply:
Done